domain/src/main/java/org/oppia/domain/oppialogger/logstorage/LogReportStorageHelper.kt

```diff
@@ -0,0 +1,175 @@
+import org.oppia.util.logging.ConsoleLogger
+import org.oppia.util.threading.BackgroundDispatcher
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Helper class for storing log reports in absence of network connectivity. */
+@Singleton
+class LogReportStorageHelper @Inject constructor(
+  private val cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val consoleLogger: ConsoleLogger,
+  @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  @EventLogStorageCacheSize private val eventLogStorageCacheSize: Int,
+  @ExceptionLogStorageCacheSize private val exceptionLogStorageCacheSize: Int
+) {
+  private enum class LogReportingCase {
+    EVENT_LOG,
+    EXCEPTION_LOG
+  }
+  private val eventLogStore =
+    cacheStoreFactory.create("event_logs", OppiaEventLogs.getDefaultInstance())
+  private val exceptionLogStore =
+    cacheStoreFactory.create("exception_logs", OppiaExceptionLogs.getDefaultInstance())
+  private val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+
+  /** Adds an event to the storage. */
+  fun addEventLog(eventLog: EventLog) {
+    coroutineScope.launch {
+      checkStoreCacheStatus(
+        eventLogStore.readDataAsync().await().eventLogList.size + 1,
+        LogReportingCase.EVENT_LOG,
+        eventLogStorageCacheSize
+      )
+    }
+    eventLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addEventLog(eventLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          "DOMAIN",
+          "Failed to store event log",
+          it
+        )
+      }
+    }
+  }
+
+  /** Adds an exception to the storage. */
+  fun addExceptionLog(exceptionLog: ExceptionLog) {
+    coroutineScope.launch {
+      checkStoreCacheStatus(
+        exceptionLogStore.readDataAsync().await().exceptionLogList.size + 1,
+        LogReportingCase.EXCEPTION_LOG,
+        exceptionLogStorageCacheSize
+      )
+    }
+    exceptionLogStore.storeDataAsync(updateInMemoryCache = true) {
+      it.toBuilder().addExceptionLog(exceptionLog).build()
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          "DOMAIN",
+          "Failed to store crash log",
+          it
+        )
+      }
+    }
+  }
+
+  /** Checks the [storeSize] and removes an element from the corresponding store if the [cacheStorageLimit] is exceeded. */
+  private suspend fun checkStoreCacheStatus(
+    storeSize: Int,
+    logReportCase: LogReportingCase,
+    cacheStorageLimit: Int
+  ) {
+    if (storeSize > cacheStorageLimit) {
+      val recordsToBeRemoved = storeSize - cacheStorageLimit
+      for (i in 1..recordsToBeRemoved) {
+        when (logReportCase) {
+          LogReportingCase.EVENT_LOG -> removeEvent(getLeastRecentEvent())
+          LogReportingCase.EXCEPTION_LOG -> removeException(getLeastRecentException())
+        }
+      }
+    }
```

Comment:
Can we instead remove all N events in one shot rather than running it in a for loop

Reply:
Actually to keep it simple lets just do one event for now (i.e. remove the for loop). If we find it to be a problem we can do remove N records

Reply:
Okay, done.